plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,284 @@
+returnL (FrameUnwrap              : ctx) heap (Closure (Wrap _ _ _ t) env) = returnL ctx heap (Closure t env)
+returnL (FrameUnwrap              : _)      _ (Closure term _)             = throw $ LMachineException "Attemtping to unwrap non-wrapped term" term
+
+
+-- | Apply a function to an argument and proceed.
+-- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+
+evaluateFun :: EvaluationContext -> Heap -> Closure -> Closure-> LMachineResult
+evaluateFun ctx heap (Closure fun funEnv) argClosure =
+    case fun of
+      LamAbs _ var _ body ->
+          let (l, heap') = insertInHeap argClosure heap
+              env' = updateEnvironment (unUnique $ nameUnique var) l funEnv
+          in  computeL ctx heap' (Closure body env')
+
+      _ ->
+          -- Not a lambda: look for evaluation of a built-in function, possibly with some args already supplied.
+          -- We have to force the arguments, which means that we need to get the new heap back as well.
+          -- This bit is messy but should be much easier when we have n-ary application for builtins.
+          case computeL [] heap argClosure of
+            -- Force the argument, but only at the top level.
+            -- This is a bit of a hack.  We're not in the main compute/return process here.
+            -- We're preserving ctx, the context at entry, and we'll use that when we return.
+            -- We could also add a new type of stack frame for this.  Exactly what we do will
+            -- depend on the final interface to builtins.
+            Failure -> Failure
+            Success (Closure arg' env') heap' ->
+                let term = Apply () fun arg'
+                in case termAsPrimIterApp term of
+                     Nothing ->
+                         throw $ LMachineException "Trying to apply invalid term" term
+                         -- Was "Cannot reduce a not immediately reducible application."  This message isn't very helpful.
+                     Just (IterApp (StaticStagedBuiltinName name) spine) ->
+                         case applyEvaluateBuiltinName heap' env' name spine of
+                           ConstAppSuccess term' -> returnL ctx heap' (Closure term' funEnv)
+                           ConstAppStuck         -> returnL ctx heap' (Closure term  funEnv)
+                           -- It's arguable what the env should be here. That depends on what the built-in can return.
+                           -- Ideally it'd always return a closed term, so the environment should be irrelevant.
```

Comment:
I'm not sure I understand this comment. Why should the `env` change here?
I think it's safe to assume that built-ins return closed terms. We may want to revisit this later for some creepy reason, but let's not bother with this right now.